File: leetcode/1178/1178.go
import (
    "math/bits"
)

func findNumOfValidWords(words []string, puzzles []string) []int {
    m := make(map[int]int)
    for _, word := range words{
        num := 0
        for _, c := range(word) {
            num |= 1<<(c - 'a')
        }
        if bits.OnesCount(uint(num)) <= 7 {
            m[num]++
        }
    }
    res := make([]int, len(puzzles))
    for idx, puzzle := range(puzzles) {
        count := 0
        for i:=0; i<(1<<6); i++ {
            temp := 1 << (puzzle[0] - 'a')
            for j:=0; j<6; j++ {
                temp |= ((i>>j)&1)<<(puzzle[j+1] - 'a')
            }
            count += m[temp]
        }
        res[idx] = count
    }
    return res
}

File: leetcode/1178/1178.py
class Solution:
    def findNumOfValidWords(self, words: List[str], puzzles: List[str]) -> List[int]:
        def w2n(s):
            num = 0
            for c in s:
                num |= 1<<(ord(c) - ord('a'))
            return num

        m = {}
        for word in words:
            if len(ws := set(word)) <= 7:
                w = w2n(ws)
                m[w] = m.get(w, 0) + 1

        def getRes(puzzle):
            count = 0
            lst = [ord(c) - ord('a') for c in puzzle]
            # 枚举所有可能的谜面
            for i in range(1<<6):
                temp = 1 << lst[0]
                for j in range(6):
                    temp |= (((i>>j) & 1) << lst[j+1])
                count += m.get(temp, 0)
            return count
        return [getRes(puzzle) for puzzle in puzzles]
